config.org
#+title: Damian's personal GNU Emacs config
#+author: Damian Suarez
#+description: now in org mode!!
#+startup: showeverything
#+options: toc:2

* Table of contents :toc:
- [[#priorities][Priorities]]
  - [[#package-manager-straightel][Package Manager (straight.el)]]
  - [[#evil-mode-][Evil mode >:)]]
  - [[#all-the-icons][All the icons]]
  - [[#font-stuff][Font stuff]]
  - [[#lsp][LSP]]
  - [[#gui][GUI]]
- [[#general-keybindings][General Keybindings]]
  - [[#space][Space]]
  - [[#evil-special-keybinds][Evil special keybinds]]
- [[#configurations][Configurations]]
  - [[#dired][Dired]]
  - [[#shells][Shells]]
- [[#more-plugins][More Plugins!!]]
  - [[#dashboard][Dashboard]]
  - [[#ivy][Ivy]]
  - [[#mood-line][Mood-line]]
  - [[#magit][Magit]]
  - [[#language-support][Language Support]]
  - [[#org-mode][Org-Mode]]
  - [[#projectile][Projectile]]
  - [[#rainbow-mode][Rainbow mode]]
  - [[#sudo-edit][Sudo-edit]]
  - [[#which-key][Which-key]]

* Priorities
** Package Manager (straight.el)
I also added early-init.el to stop package.el from loading, as straight is more flexible

#+BEGIN_SRC emacs-lisp
(defvar bootstrap-version)
(let ((bootstrap-file
    (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
    (bootstrap-version 6))
(unless (file-exists-p bootstrap-file)
        (with-current-buffer
        (url-retrieve-synchronously
        "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
        'silent 'inhibit-cookies)
    (goto-char (point-max))
    (eval-print-last-sexp)))
(load bootstrap-file nil 'nomessage))
;; use-package macro and set to default
(straight-use-package 'use-package)
(setq straight-use-package-by-default t)
#+END_SRC

** Evil mode >:)
vi-vi-vi is the symbol of the beast!
#+BEGIN_SRC emacs-lisp
(use-package evil
   :init
   (setq evil-want-integration t)
   (setq evil-want-keybinding nil)
   (setq evil-want-vsplit-window-right t)
   (setq evil-want-split-window-below t)
   (evil-mode))

(use-package evil-collection
   :after evil
   :config
   (setq evil-collection-mode-list '(org dashboard dired ibuffer magit))
   (evil-collection-init))

(use-package evil-commentary)
(evil-commentary-mode)
#+END_SRC

Why isn't this a default?
#+begin_src emacs-lisp
(global-set-key (kbd "C-u") 'evil-scroll-up)
#+end_src

** All the icons
I love them all
#+begin_src emacs-lisp
(use-package all-the-icons
:ensure t
:if (display-graphic-p))

(use-package all-the-icons-dired
:hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))
#+end_src

** Font stuff
Because I really like fira
*** Default Fonts
#+BEGIN_SRC emacs-lisp
(set-face-attribute 'default nil
  :font "FiraCode Nerd Font"
  :height 140
  :weight 'medium)
(set-face-attribute 'variable-pitch nil
  :font "Source Code Pro"
  :height 140
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "FiraCode Nerd Font"
  :height 140
  :weight 'medium)
#+END_SRC

*** Italics
#+BEGIN_SRC emacs-lisp
  (set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)
#+END_SRC

*** Zoom
and enhance
#+begin_src emacs-lisp
(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
#+end_src

*** Client config and more!
Because client is wonky!!
#+BEGIN_SRC emacs-lisp
(add-to-list 'default-frame-alist '(font . "FiraCode Nerd Font-14"))
(setq-default line-spacing 0.12)
#+END_SRC

** LSP
M-x vscode
#+begin_src emacs-lisp
(use-package lsp-mode
  :init
  (setq lsp-keymap-prefix "C-c l")
  :hook ((XXX-mode . lsp)
         (lsp-mode . lsp-enable-which-key-integration))
  :commands lsp)

;; optionally
(use-package lsp-ui :commands lsp-ui-mode)
(use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
(use-package dap-mode)
(use-package which-key
    :config
    (which-key-mode))

(use-package company)
(use-package yasnippet-snippets)
#+end_src

** GUI
Because more is not always better

*** Disable GUI stuff
A menu? in MY emacs??
#+BEGIN_SRC emacs-lisp
  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1)
#+END_SRC

*** Display Line Numbers
relative numbers or bust
#+BEGIN_SRC emacs-lisp
  (global-display-line-numbers-mode 1)
  (global-visual-line-mode t)
  (setq display-line-numbers-type 'relative)
#+END_SRC

*** Smooth scrolling
who needs C-v
#+begin_src emacs-lisp
(setq scroll-step            1
      scroll-conservatively  10000
      scroll-margin 8)
#+end_src

*** Colorscheme
spaceduck is installed now!!! pick your poison
uncomment this block and comment the next or viceversa
**** Catpuccin
catppuccin is fun
#+BEGIN_SRC emacs-lisp
   (add-to-list 'custom-theme-load-path "~/.config/emacs/themes/")
   (straight-use-package 'catppuccin-theme)
   (load-theme 'catppuccin :no-confirm)
  (setq catppuccin-flavor 'mocha) ;; or 'frappe, 'latte, or 'mocha
  ;; (catppuccin-reload)
#+END_SRC

**** Personal Theme 
it's called bloom :)
#+BEGIN_SRC emacs-lisp
;;(load-theme 'Bloom t)
#+END_SRC

**** Spaceduck
A duck?
#+BEGIN_SRC emacs-lisp
;;(load-theme 'spaceduck t)
#+END_SRC

* General Keybindings
** Space
other keybinds may be on their specific configs
#+BEGIN_SRC emacs-lisp
  (use-package general
      :config
      (general-evil-setup)

  ;; set up Space as the global leader key
  (general-create-definer leader-keys
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ;; set leader
      :global-prefix "M-SPC")

  ;; File
  ;; TODO: highlighted 
  (leader-keys
      "f" '(:ignore :wk "File")
      "." '(find-file :wk "Find file")
      "fr" '(counsel-recentf :wk "Find recent file")
      "fp" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit config")
      "fs" '(save-buffer :wk "Save File"))

  ;; Buffer
  (leader-keys
      "b" '(:ignore t :wk "Buffer")
      "bb" '(switch-to-buffer :wk "Switch buffer")
      "bi" '(ibuffer :wk "Ibuffer")
      "bd" '(kill-this-buffer :wk "Kill this buffer")
      "bn" '(next-buffer :wk "Next buffer")
      "bp" '(previous-buffer :wk "Previous buffer")
      "br" '(revert-buffer :wk "Reload buffer"))

  ;; Elisp eval 
  (leader-keys
      "e" '(:ignore t :wk "Evaluate")
      "eb" '(eval-buffer :wk "Eval elisp in buffer")
      "ed" '(eval-defun :wk "Eval defun containing or after point")
      "ee" '(eval-expression :wk "Eval an elisp expression")
      "el" '(eval-last-sexp :wk "Eval elisp expression before point")
      "er" '(eval-region :wk "Eval elisp in region"))

  ;; Help (I need somebody :o)
  (leader-keys
      "h" '(:ignore t :wk "Help")
      "hf" '(describe-function :wk "Describe function")
      "hv" '(describe-variable :wk "Describe variable")
      "hrr" '((lambda () (interactive) (load-file "~/.config/emacs/init.el")) :wk "Reload emacs config"))

  ;; Window management
  (leader-keys
      "w" '(:ignore t :wk "Window")
      "wh" '(evil-window-left :wk "Window left")
      "wj" '(evil-window-down :wk "Window down")
      "wk" '(evil-window-up :wk "Window up")
      "wl" '(evil-window-right :wk "Window right")
      "wq" '(evil-window-delete :wk "Delete current window")
      "wv" '(evil-window-vsplit :wk "Move to the left window")
      "ws" '(evil-window-split :wk "Move to the left window"))

  ;; Open stuff 
  (leader-keys
      "o" '(:ignore t :wk "Open")
      "ot" '(vterm-toggle :wk "Vterm toggle")
      "op" '(treemacs :wk "treemacs"))
  )
#+END_SRC

** Evil special keybinds
#+begin_src emacs-lisp
(general-create-definer evil-keys
    :keymaps 'override
    )

(evil-keys
    :prefix "j"
    :states '(insert)
    "k" '(evil-normal-state :wk "Normal state"))
#+end_src

* Configurations
** Dired
** Shells
*** Eshell
#+begin_src emacs-lisp
(use-package eshell-syntax-highlighting
    :after  esh-mode
    :config
    (eshell-syntax-highlighting-global-mode +1))

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))
#+end_src
*** Vterm
The best terminal 
you can probably use shell-file-name but it was bugged for me :o
#+begin_src emacs-lisp
(use-package vterm
:config
(setq vterm-shell "/bin/zsh"))
#+end_src
*** Vterm-Toggle
#+begin_src emacs-lisp
(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.3))))
#+end_src

* More Plugins!!
Because that's the emacs experience

** Dashboard
Dashin' to the startup
#+begin_src emacs-lisp
(use-package dashboard
  :ensure t 
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Welcome to GNU Emacs!")
  (setq dashboard-startup-banner 'official) ;; use standard emacs logo as banner
  (setq dashboard-center-content t) ;; set to 't' for centered content
  (setq dashboard-items '((recents  . 3)
                         (bookmarks . 5)
                         (projects . 5)
                         (agenda . 5)))
  :config
  (dashboard-setup-startup-hook))

(setq dashboard-footer-messages '("TQMT"))
(setq dashboard-set-init-info nil)
#+end_src

** Ivy 
Automated lazyness
#+begin_src emacs-lisp
(use-package counsel
  :after ivy
  :config (counsel-mode))

(use-package ivy
  :bind
  ;; ivy-resume resumes the last Ivy-based completion.
  (("C-c C-r" . ivy-resume)
   ("C-x B" . ivy-switch-buffer-other-window))
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package ivy-rich
  :after ivy
  :ensure t
  :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
  :custom
  (ivy-virtual-abbreviate 'full
   ivy-rich-switch-buffer-align-virtual-buffer t
   ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
                               'ivy-rich-switch-buffer-transformer))
#+end_src

** Mood-line
#+begin_src emacs-lisp
(use-package nerd-icons)
(use-package doom-modeline)
(doom-modeline-mode 1)
#+end_src

** Magit
git has never been easier
*** Installation and config
#+begin_src emacs-lisp
(use-package magit)
(setq magit-display-buffer-function #'magit-display-buffer-fullframe-status-v1)
#+end_src

*** Keybinds
#+begin_src emacs-lisp
(leader-keys
    "g" '(:ignore t :wk "Git")
    "gg" '(magit-status :wk "Magit Status"))
#+end_src

** Language Support
*** Languages
Emacs has built-in programming language modes for Lisp, Scheme, DSSSL, Ada, ASM, AWK, C, C++, Fortran, Icon, IDL (CORBA), IDLWAVE, Java, Javascript, M4, Makefiles, Metafont, Modula2, Object Pascal, Objective-C, Octave, Pascal, Perl, Pike, PostScript, Prolog, Python, Ruby, Simula, SQL, Tcl, Verilog, and VHDL.  Other languages will require you to install additional modes.
#+begin_src emacs-lisp
(use-package go-mode)
(use-package lua-mode)
(use-package web-mode)
#+end_src

*** Flycheck
#+begin_src emacs-lisp
(use-package flycheck
  :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))
#+end_src
*** Diminish
#+begin_src emacs-lisp
(use-package diminish)
#+end_src
** Org-Mode
Making life simple and avoiding unorganized crime
*** Enabling table of contents
#+BEGIN_SRC emacs-lisp
(use-package toc-org
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))
#+END_SRC

*** Folder
#+begin_src emacs-lisp
(setq org-agenda-files (quote ("~/Agenda/")))
#+end_src
*** Org-bullets
pew pew
#+BEGIN_SRC emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+END_SRC

*** Weird TAB behaviour 
weird org block indents fix :)
#+BEGIN_SRC emacs-lisp
(electric-indent-mode -1)
(setq org-src-preserve-indentation t)
(setq ation t)
#+END_SRC

*** Org-tempo
like emmet but good
#+BEGIN_SRC emacs-lisp
(require 'org-tempo)
#+END_SRC

*** Keybinds
#+begin_src emacs-lisp
(leader-keys
    "m" '(:ignore t :wk "Org")
    "oa" '(org-agenda :wk "Org agenda")
    "me" '(org-export-dispatch :wk "Org export dispatch")
    "mi" '(org-toggle-item :wk "Org toggle item")
    "mt" '(org-todo :wk "Org todo")
    "mB" '(org-babel-tangle :wk "Org babel tangle")
    "mT" '(org-todo-list :wk "Org todo list"))

(leader-keys
    "mb" '(:ignore t :wk "Tables")
    "mb -" '(org-table-insert-hline :wk "Insert hline in table"))

(leader-keys
    "md" '(:ignore t :wk "Date/deadline")
    "mdt" '(org-time-stamp :wk "Org time stamp"))

(leader-keys
    "p" '(projectile-command-map :wk "Projectile")
    "ff" '(projectile-find-file :wk "Projectile find file"))
#+end_src

** Projectile
To aim high with your projects
#+begin_src emacs-lisp
(use-package projectile
    :config
    (projectile-mode 1))
#+end_src
** Rainbow mode
pride flags galore
#B00B69
#A55EA7
#1D1CC9 
#+begin_src emacs-lisp
(use-package rainbow-mode
    :hook org-mode prog-mode)
#+end_src
 
** Sudo-edit
what's the magic word to get what you want?
#+begin_src emacs-lisp
(use-package sudo-edit
    :config
    (leader-keys
        "fu" '(sudo-edit-find-file :wk "Sudo find file")
        "fU" '(sudo-edit :wk "Sudo edit this file")))
#+end_src

** Which-key
you expect me to use sane keybinds??
#+BEGIN_SRC emacs-lisp
(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	which-key-sort-order #'which-key-key-order-alpha
	which-key-sort-uppercase-first nil
	which-key-add-column-padding 1
	which-key-display-columns nil
	which-key-min-display-lines 6
	which-key-side-window-slot -10
	which-key-side-window-max-height 0.25
	which-key-idle-delay 0.8
	which-key-max-description-length 25
	which-key-allow-imprecise-window-fit nil
	which-key-separator " → "))
#+END_SRC


config.el
(defvar bootstrap-version)
(let ((bootstrap-file
    (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
    (bootstrap-version 6))
(unless (file-exists-p bootstrap-file)
        (with-current-buffer
        (url-retrieve-synchronously
        "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
        'silent 'inhibit-cookies)
    (goto-char (point-max))
    (eval-print-last-sexp)))
(load bootstrap-file nil 'nomessage))
;; use-package macro and set to default
(straight-use-package 'use-package)
(setq straight-use-package-by-default t)

(use-package evil
   :init
   (setq evil-want-integration t)
   (setq evil-want-keybinding nil)
   (setq evil-want-vsplit-window-right t)
   (setq evil-want-split-window-below t)
   (evil-mode))

(use-package evil-collection
   :after evil
   :config
   (setq evil-collection-mode-list '(org dashboard dired ibuffer magit))
   (evil-collection-init))

(use-package evil-commentary)
(evil-commentary-mode)

(global-set-key (kbd "C-u") 'evil-scroll-up)

(use-package all-the-icons
:ensure t
:if (display-graphic-p))

(use-package all-the-icons-dired
:hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))

(set-face-attribute 'default nil
  :font "FiraCode Nerd Font"
  :height 140
  :weight 'medium)
(set-face-attribute 'variable-pitch nil
  :font "Source Code Pro"
  :height 140
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "FiraCode Nerd Font"
  :height 140
  :weight 'medium)

  (set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(add-to-list 'default-frame-alist '(font . "FiraCode Nerd Font-14"))
(setq-default line-spacing 0.12)

(use-package lsp-mode
  :init
  (setq lsp-keymap-prefix "C-c l")
  :hook ((XXX-mode . lsp)
         (lsp-mode . lsp-enable-which-key-integration))
  :commands lsp)

;; optionally
(use-package lsp-ui :commands lsp-ui-mode)
(use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
(use-package dap-mode)
(use-package which-key
    :config
    (which-key-mode))

(use-package company)
(use-package yasnippet-snippets)

  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1)

  (global-display-line-numbers-mode 1)
  (global-visual-line-mode t)
  (setq display-line-numbers-type 'relative)

(setq scroll-step            1
      scroll-conservatively  10000
      scroll-margin 8)

   (add-to-list 'custom-theme-load-path "~/.config/emacs/themes/")
   (straight-use-package 'catppuccin-theme)
   (load-theme 'catppuccin :no-confirm)
  (setq catppuccin-flavor 'mocha) ;; or 'frappe, 'latte, or 'mocha
  ;; (catppuccin-reload)

;;(load-theme 'Bloom t)

;;(load-theme 'spaceduck t)

  (use-package general
      :config
      (general-evil-setup)

  ;; set up Space as the global leader key
  (general-create-definer leader-keys
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ;; set leader
      :global-prefix "M-SPC")

  ;; File
  ;; TODO: highlighted 
  (leader-keys
      "f" '(:ignore :wk "File")
      "." '(find-file :wk "Find file")
      "fr" '(counsel-recentf :wk "Find recent file")
      "fp" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit config")
      "fs" '(save-buffer :wk "Save File"))

  ;; Buffer
  (leader-keys
      "b" '(:ignore t :wk "Buffer")
      "bb" '(switch-to-buffer :wk "Switch buffer")
      "bi" '(ibuffer :wk "Ibuffer")
      "bd" '(kill-this-buffer :wk "Kill this buffer")
      "bn" '(next-buffer :wk "Next buffer")
      "bp" '(previous-buffer :wk "Previous buffer")
      "br" '(revert-buffer :wk "Reload buffer"))

  ;; Elisp eval 
  (leader-keys
      "e" '(:ignore t :wk "Evaluate")
      "eb" '(eval-buffer :wk "Eval elisp in buffer")
      "ed" '(eval-defun :wk "Eval defun containing or after point")
      "ee" '(eval-expression :wk "Eval an elisp expression")
      "el" '(eval-last-sexp :wk "Eval elisp expression before point")
      "er" '(eval-region :wk "Eval elisp in region"))

  ;; Help (I need somebody :o)
  (leader-keys
      "h" '(:ignore t :wk "Help")
      "hf" '(describe-function :wk "Describe function")
      "hv" '(describe-variable :wk "Describe variable")
      "hrr" '((lambda () (interactive) (load-file "~/.config/emacs/init.el")) :wk "Reload emacs config"))

  ;; Window management
  (leader-keys
      "w" '(:ignore t :wk "Window")
      "wh" '(evil-window-left :wk "Window left")
      "wj" '(evil-window-down :wk "Window down")
      "wk" '(evil-window-up :wk "Window up")
      "wl" '(evil-window-right :wk "Window right")
      "wq" '(evil-window-delete :wk "Delete current window")
      "wv" '(evil-window-vsplit :wk "Move to the left window")
      "ws" '(evil-window-split :wk "Move to the left window"))

  ;; Open stuff 
  (leader-keys
      "o" '(:ignore t :wk "Open")
      "ot" '(vterm-toggle :wk "Vterm toggle")
      "op" '(treemacs :wk "treemacs"))
  )

(general-create-definer evil-keys
    :keymaps 'override
    )

(evil-keys
    :prefix "j"
    :states '(insert)
    "k" '(evil-normal-state :wk "Normal state"))

(use-package eshell-syntax-highlighting
    :after  esh-mode
    :config
    (eshell-syntax-highlighting-global-mode +1))

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))

(use-package vterm
:config
(setq vterm-shell "/bin/zsh"))

(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.3))))

(use-package dashboard
  :ensure t 
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Welcome to GNU Emacs!")
  (setq dashboard-startup-banner 'official) ;; use standard emacs logo as banner
  (setq dashboard-center-content t) ;; set to 't' for centered content
  (setq dashboard-items '((recents  . 3)
                         (bookmarks . 5)
                         (projects . 5)
                         (agenda . 5)))
  :config
  (dashboard-setup-startup-hook))

(setq dashboard-footer-messages '("TQMT"))
(setq dashboard-set-init-info nil)

(use-package counsel
  :after ivy
  :config (counsel-mode))

(use-package ivy
  :bind
  ;; ivy-resume resumes the last Ivy-based completion.
  (("C-c C-r" . ivy-resume)
   ("C-x B" . ivy-switch-buffer-other-window))
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package ivy-rich
  :after ivy
  :ensure t
  :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
  :custom
  (ivy-virtual-abbreviate 'full
   ivy-rich-switch-buffer-align-virtual-buffer t
   ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
                               'ivy-rich-switch-buffer-transformer))

(use-package nerd-icons)
(use-package doom-modeline)
(doom-modeline-mode 1)

(use-package magit)
(setq magit-display-buffer-function #'magit-display-buffer-fullframe-status-v1)

(leader-keys
    "g" '(:ignore t :wk "Git")
    "gg" '(magit-status :wk "Magit Status"))

(use-package go-mode)
(use-package lua-mode)
(use-package web-mode)

(use-package flycheck
  :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))

(use-package diminish)

(use-package toc-org
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))

(setq org-agenda-files (quote ("~/Agenda/")))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(electric-indent-mode -1)
(setq org-src-preserve-indentation t)
(setq ation t)

(require 'org-tempo)

(leader-keys
    "m" '(:ignore t :wk "Org")
    "oa" '(org-agenda :wk "Org agenda")
    "me" '(org-export-dispatch :wk "Org export dispatch")
    "mi" '(org-toggle-item :wk "Org toggle item")
    "mt" '(org-todo :wk "Org todo")
    "mB" '(org-babel-tangle :wk "Org babel tangle")
    "mT" '(org-todo-list :wk "Org todo list"))

(leader-keys
    "mb" '(:ignore t :wk "Tables")
    "mb -" '(org-table-insert-hline :wk "Insert hline in table"))

(leader-keys
    "md" '(:ignore t :wk "Date/deadline")
    "mdt" '(org-time-stamp :wk "Org time stamp"))

(leader-keys
    "p" '(projectile-command-map :wk "Projectile")
    "ff" '(projectile-find-file :wk "Projectile find file"))

(use-package projectile
    :config
    (projectile-mode 1))

(use-package rainbow-mode
    :hook org-mode prog-mode)

(use-package sudo-edit
    :config
    (leader-keys
        "fu" '(sudo-edit-find-file :wk "Sudo find file")
        "fU" '(sudo-edit :wk "Sudo edit this file")))

(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	which-key-sort-order #'which-key-key-order-alpha
	which-key-sort-uppercase-first nil
	which-key-add-column-padding 1
	which-key-display-columns nil
	which-key-min-display-lines 6
	which-key-side-window-slot -10
	which-key-side-window-max-height 0.25
	which-key-idle-delay 0.8
	which-key-max-description-length 25
	which-key-allow-imprecise-window-fit nil
	which-key-separator " → "))
